qpdf/qtest/qpdf/test-89.out

```diff
@@ -1,5 +1,5 @@
-WARNING: manual-qpdf-json.json, trailer at offset 1801: operation for array attempted on object of type dictionary: ignoring attempt to append item
-WARNING: manual-qpdf-json.json, obj:1 0 R at offset 1119: operation for array attempted on object of type dictionary: ignoring attempt to append item
```

Comment:
`obj:1 0 R` is the name of the key in the json file. `object 1 0` is what it represents, but I'd rather see the actual key in the error message.

Reply:
That is the "slightly less information in error messages" I mentioned.

The trade-offs are:

- Construct individual object descriptions and add 15% to the runtime (i.e. forgo the 13% potential saving on the status quo) for runs that use a JSON input file.
- Store the key with the object and use a template string to construct the warning. Adding an extra string to QPDFValue adds somewhere between 0.5% to 1% to the runtime of QPDF, whether it uses JSON input or not.
- Drop the key and rely on the offset (and maybe theObjGen which we get for free), the proposed solution.

In addition to the 3 simple options above there are 2 further options that don't add any real overhead unless warnings are issued:

- Instead of storing a shared pointer to a template string with QPDFValues, have a shared pointer to a struct containing (at least) the input source and the template string. Calculate the key from the input source and offset only when needed.
- Provide a cmdline option to provide additional info given the JSON input file and offset.

To me the last 3 options are the most attractive, and the proposed option is probably sufficient. This is based on:
- I prefer not to have to pay unless I use it.
- Given that JSON is a pure text file, the offset is probably sufficient to locate the problem . (Most users who receive warnings probably won't be interested)
- Offset and key are probably insufficient info for those users who are interested. Additional info that what be useful:
   - full path of the actual object that triggered the warning
   - line no  and col corresponding to the offset, and possibly:
   - offset, line no and col for each element of the path, and
   - the actual line, surrounding lines, and marker highlighting the point of problem.

Reply:
Its probably useful to be clear about the audience for the warnings - are they for the benefit of end users or of developers tracking down programming errors. The two groups have different needs. Arguably QPDF provides too much info for the first group and two little for the second.

Reply:
They are for end users who are trying to create or use qpdf json files. I don't think they contain too much information. I don't want to sacrifice error messages for performance. I looked at it a little too fast and was thinking these were json errors, not qpdf errors. I might be okay with `obj:1 0 R` becoming `object 1 0` if there were a really compelling advantage, but I'm not okay with `trailer` becoming `object 0` or `object 0 0`. `trailer` has to be `trailer`. `object 0` looks like a mistake. But I'd rather have it be `obj:1 0 R` because it will be more obvious to someone who created a json with errors in it that they are in the right place, and I don't think a 13% improvement in json parsing is going to move the needle for most qpdf users.

Adding a command line to make warnings more specific is out. The command-line is already too complicated, and this won't help when other tools consume qpdf json.

Adding context for errors is too much. A byte offset is fine. qpdf is not a compiler.

I don't think we can sacrifice usability in the name of performance. We're talking about a linear improvement on code path that is not run repeatedly. Basically you're creating these object descriptions when the file is parsed. We could clearly gain a lot of performance by dropping object descriptions entirely, but it makes it much harder to work with error messages.

Basically I don't see the performance of qpdf json as a problem that requires other sacrifices to solve. Things that can be improved through simple refactoring, like checking token type codes instead of using dynamic cast or other optimizations to the parser are great -- they have an upside with no real downside. Squeezing out extra performance at the cost of user confusion when interpreting error messages feels like a wrong trade to me.

Reply:
Whether the error message says ```obj:1 0 R``` or ```object 1 0``` is not an issue - as long as we can pick up the obj gen from the QPDFValue rather than from the json file - which is consistent with what we are doing for objects read from a pdf file.

Having a separate template for trailer is not an issue  either - I got myself confused with messages containing a json path and thought `trailer` was just one of many possible values, which would be a problem.

I've pushed up a second attempt. If that one is acceptable, I would suggest we drop the ```obj:0 0 R``` for direct objects and simply go for "object at offset ...".